apps/events/viewsets.py

```diff
@@ -102,16 +102,29 @@ def get_template_names(self):
 
     def get_queryset(self):
         if self.kwargs.get("pk"):
-            self.queryset = Event.objects.filter(id=self.kwargs["pk"])
+            self.queryset = Event.objects.filter(
+                Q(id=self.kwargs["pk"]) & Q(is_visible=True) & Q(is_finished=False) & Q(type="open") |
+                Q(id=self.kwargs["pk"]) & Q(participants__in=[self.request.user.id]) & Q(is_finished=False) |
+                Q(id=self.kwargs["pk"]) & Q(created_by=self.request.user) & Q(type="private") |
+                Q(id=self.kwargs["pk"]) & Q(type="private") & Q(is_finished=True) & Q(is_visible=True)
+            )
+        if self.kwargs.get("token"):
+            self.queryset = Event.objects.filter(
+                Q(private_url=self.kwargs["token"]) & Q(created_by=self.request.user) |
+                Q(private_url=self.kwargs["token"] & Q(is_finished=False))
+            )
```

Comment:
We can remove this if we implement get_object with queryset parameter in action

Reply:
if self.kwargs.get("token"):
            self.queryset = Event.objects.filter(private_url=self.kwargs["token"])

Reply:
get_object  didn't override